### DISCUSSION TITLE
Why does TPE algorithm only work for a while and then revert to random sampling?

### DISCUSSION THREAD

**Discussion by mayushii21** [optuna/optuna#4654] [Q&A] [answered]
Could someone please explain why it seems like the TPE algorithm only works for a while and then reverts back to random sampling?
![newplot (7)](https://user-images.githubusercontent.com/122178787/236622231-7b448633-c7d6-4927-935e-330bef3d7486.png)
At first I thought it might have to do with the algorithm not having "enough" data about the search space, so I thought it might be doing some random sampling to "broaden its view", so I tried increasing the n_startup_trials parameter. But that only moved the reversion to what seems like random sampling further down the road.
![newplot (6)](https://user-images.githubusercontent.com/122178787/236622265-8d4d9a89-e2af-435c-9afd-4d4041843364.png)
Then I noticed that there seems to be roughly 200 "smart" trials after the n_startup_trials, so I assumed that maybe the TPE algorithm only works for 200 trials, concludes that it "converged" to the optimal solution, and then just tries randomly searching for a better option, so I tried lowering the total number of trials to n_startup_trials+200.
![newplot (8)](https://user-images.githubusercontent.com/122178787/236622416-1c6379fa-ea03-49b9-b021-098f01146cc2.png)
But here again, it seems, that the "smart" algorithm works for a while, but then starts sampling randomly after a while..
I just don't understand this behavior.. is it supposed to be like this? Why? Could someone please explain it in simple terms?

The way I see it, it makes little sense to continue (dark blue is newer trials) to sample low values for the eta parameter, if you can clearly see that it consistently gives poor results..
![Screenshot 2023-05-06 155736](https://user-images.githubusercontent.com/122178787/236622747-95bb56b9-16b4-4108-bad0-bae5748e2745.png)

**Comment by contramundum53** [comment, marked as answer]
Intuitively, TPE algorithm works in the following style:
* First it samples randomly.
* Then it gradually shifts the distribution towards "good parameters" and away from "bad parameters". Here, "good parameters" are by default the smaller of top 10% or 25 trials and "bad parameters" are all the other trials. (The number of "good parameters" can be adjusted by the parameter `gamma` in `TPESampler`.) Thus, hopefully, better and better parameters are sampled as the optimization proceeds. On the other hands, if there are too many "bad parameters" around "good parameters", it starts looking for other sweet spots that may contain good parameters.
* As the optimization proceeds, the distribution becomes sharper and sharper and finally converges towards the optimal parameter (optimal in a specific sense, since the objective function is in general stochastic). However, the rate of the distribution shift is limited by "magic clip", which prevents premature exploitation when the search space is large. Thus it will still sample small `eta`s with some probability while the distribution is not sharp enough. Unfortunately, we do not currently have interface to adjust the rate of magic clip for now.

In conclusion, if the search space is relatively small and the number of trials is relatively large, the algorithm deems that "it has explored enough parameters around current best parameters for now" and starts trying its luck in different regions. (Also note that TPE does not see "how bad" the previous parameters are; all parameters that do not rank in as the top 25 are "equally bad".)

If `eta` is the only parameter you want to optimize and the objective is smooth like your figure, I would recommend you use `optuna.integration.botorch.BoTorchSampler`. I think it would give you pretty precise solution within several hundred evaluations. On the other hand, if you want to have >1000 trials to find very precise optimal solution, I recommend you use `CmaEsSampler`.

**Reply by mayushii21** [reply]
Thank you for the reply. I didn't know about `BoTorchSampler` (tried `SkoptSampler` but it turned out to be very slow and practically unusable with 500 or more trials). Does it support categorical distributions?
What if I have many (7-15), both continuous and discrete parameters? Would those work well for the job? I tried `CmaEsSampler`, but couldn't figure out how margin works and what it does.. when I enabled it, it seemed to me like discrete parameters were treated as min/max binary choices (they were mostly sampled only at the top and bottom of the given range; increasing the range didn't change anything).. I do like the results of the `CmaEsSampler` for large numbers of trials tho, but it doesn't support categorical parameters, which is an inconvenience.
On a sidenote, does `CmaEsSampler` benefit from an increased n_startup_trials? It's default is 1 (opposed to 10 for `TPESampler` and, as I see, `BoTorchSampler` as well). Would increasing the number of initial random samplings help avoid getting stuck in a local optima and/or possibly speed up convergence when many parameters are being optimized?
Thank you for your time!

**Reply by contramundum53** [reply]
Currently BoTorchSampler works very poorly for categorical variables due to the difficulty in optimizing acquisition function in mixed-integer space. 
CMA-ES with margin acting like min/max binary choices is a bug that is fixed in Optuna v3.2, so please update the optuna version.
CMA-ES sampler does not benefit from increased n_startup_trials. If you have local-optima problems, you can enable restart options of CMA-ES sampler.

**Reply by contramundum53** [reply]
BoTorchSampler is slow for >1000 trials either. It's an inherent problem existing in Gaussian Process-based optimization algorithms.

**Comment by mayushii21** [comment]
thank you for the replies!
